### PR TITLE
feat(protocol): SingleBatch Type

### PR DESCRIPTION
**Pull request by refcell**
### Description

Part of a port migrating the batch types from `kona-derive` to `op-alloy`. See [`kona#695`](https://github.com/anton-rs/kona/issues/695).

This PR introduces the `SingleBatch` type.

### CODE REVIEW

**Comment by refcell** [comment]
## 📚 $\text{Stack Overview}$

Pulls submitted in this stack:
* #188 👈
* #187
* #186
* `main`

_This comment was automatically generated by [`st`](https://github.com/clabby/st)._

File: crates/protocol/src/batch/single.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            return if cfg.is_holocene_active(inclusion_block.timestamp) {
+                BatchValidity::Past
+            } else {
+                BatchValidity::Drop
+            };
+        }
+        BatchValidity::Accept
+    }
+
+    /// Checks if the batch is valid.
```
Comment (onbjerg) [inline crates/protocol/src/batch/single.rs:61]: nit: would be great to link to the specs if there are up to date ones

File: crates/protocol/src/batch/single.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                return BatchValidity::Drop;
+            }
+            return BatchValidity::Future;
+        }
+        if self.timestamp < next_timestamp {
+            return if cfg.is_holocene_active(inclusion_block.timestamp) {
+                BatchValidity::Past
+            } else {
+                BatchValidity::Drop
+            };
```
Comment (onbjerg) [inline crates/protocol/src/batch/single.rs:56]: nit: can we pick either `return if` or `if .. { return }`

**Review by onbjerg — APPROVED** [review]
otherwise lgtm

File: crates/protocol/src/batch/single.rs (hunk 1 above)
Comment (refcell) [reply crates/protocol/src/batch/single.rs:61]: Great suggestion.

**Pull request merged**